elide-datastore/elide-datastore-hibernate/src/main/java/com/yahoo/elide/core/hibernate/hql/AbstractHQLQueryBuilder.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright 2017, Oath Inc.
+ * Licensed under the Apache License, Version 2.0
+ * See LICENSE file in project root for terms.
+ */
+package com.yahoo.elide.core.hibernate.hql;
+
+import com.yahoo.elide.core.EntityDictionary;
+import com.yahoo.elide.core.filter.FilterPredicate;
+import com.yahoo.elide.core.filter.expression.FilterExpression;
+import com.yahoo.elide.core.filter.expression.PredicateExtractionVisitor;
+import com.yahoo.elide.core.hibernate.Query;
+import com.yahoo.elide.core.hibernate.Session;
+import com.yahoo.elide.core.pagination.Pagination;
+import com.yahoo.elide.core.sort.Sorting;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Abstract class used to construct HQL queries.
+ */
+public abstract class AbstractHQLQueryBuilder {
+    protected final Session session;
+    protected final EntityDictionary dictionary;
+
+    protected Optional<Sorting> sorting;
+    protected Optional<Pagination> pagination;
+    protected Optional<FilterExpression> filterExpression;
+    protected static final String SPACE = " ";
+    protected static final String UNDERSCORE = "_";
+    protected static final String PERIOD = ".";
+    protected static final String COMMA = ",";
+    protected static final String FROM = " FROM ";
+    protected static final String JOIN = " JOIN ";
+    protected static final String SELECT = "SELECT ";
+    protected static final String AS = " AS ";
+
+    /**
+     * Represents a relationship between two entities.
+     */
+    public interface Relationship {
+        public Class<?> getParentType();
+
+        public Class<?> getChildType();
+
+        public String getRelationshipName();
+
+        public Object getParent();
+
+        public Collection getChildren();
+    }
+
+    public AbstractHQLQueryBuilder(EntityDictionary dictionary, Session session) {
+        this.session = session;
+        this.dictionary = dictionary;
+
+        sorting = Optional.empty();
+        pagination = Optional.empty();
+        filterExpression = Optional.empty();
+    }
+
+    public abstract Query build();
+
+    public AbstractHQLQueryBuilder withPossibleFilterExpression(Optional<FilterExpression> filterExpression) {
+        this.filterExpression = filterExpression;
+        return this;
+    }
+
+    public AbstractHQLQueryBuilder withFilterExpression(FilterExpression filterExpression) {
+        return withPossibleFilterExpression(Optional.of(filterExpression));
```

Comment:
Given that there is another interface that takes the Optional, I think it is ok to require a non-null value here.